snorkel/slicing/modules/slice_combiner.py

```diff
@@ -44,83 +44,79 @@ def __init__(
-                if self.slice_ind_key in flow_name
-            ]
-        )
-        slice_pred_op_names = sorted(
-            [
-                flow_name
-                for flow_name in outputs.keys()
-                if self.slice_pred_key in flow_name
-            ]
+        # Concatenate indicator head predictions into tensor [batch_size x num_slices]
+        indicator_outputs = self._collect_flow_outputs_by_key(
+            flow_dict, self.slice_ind_key
         )
-
-        # Concatenate the predictions from the predictor head/indicator head
-        # into a [batch_size x num_slices] tensor
         indicator_preds = torch.cat(
             [
-                F.softmax(outputs[slice_ind_name][0])[:, 0].unsqueeze(1)
-                for slice_ind_name in slice_ind_op_names
+                F.softmax(output, dim=1)[:, 0].unsqueeze(1)
+                for output in indicator_outputs
             ],
             dim=-1,
         )
-        predictor_preds = torch.cat(
+
+        # Concatenate predictor head confidences into tensor [batch_size x num_slices]
+        predictor_outputs = self._collect_flow_outputs_by_key(
+            flow_dict, self.slice_pred_key
+        )
+
+        predictor_confidences = torch.cat(
             [
-                F.softmax(outputs[slice_pred_name][0])[:, 0].unsqueeze(1)
-                for slice_pred_name in slice_pred_op_names
+                # Compute the "confidence" using the max score across classes
+                torch.max(F.softmax(output, dim=1), dim=1)[0].unsqueeze(1)
+                for output in predictor_outputs
             ],
             dim=-1,
         )
 
-        # Collect names of predictor "features" that will be combined into the final
-        # reweighted representation
-        slice_feat_names = sorted(
-            [
-                flow_name
-                for flow_name in outputs.keys()
-                if self.slice_pred_feat_key in flow_name
-            ]
+        # Concatenate each predictor feature (to be combined into reweighted
+        # representation) into [batch_size x 1 x feat_dim] tensor
+        predictor_feat_outputs = self._collect_flow_outputs_by_key(
+            flow_dict, self.slice_pred_feat_key
         )
-
-        # Concatenate each predictor feature into [batch_size x 1 x feat_dim] tensor
         slice_representations = torch.cat(
-            [
-                outputs[slice_feat_name][0].unsqueeze(1)
-                for slice_feat_name in slice_feat_names
-            ],
-            dim=1,
+            [output.unsqueeze(1) for output in predictor_feat_outputs], dim=1
         )
 
         # Attention weights used to combine each of the slice_representations
         # incorporates the indicator (whether we are in the slice or not) and
         # predictor (confidence of a learned slice head)
-        A = F.softmax(indicator_preds * predictor_preds, dim=1)
+        A = F.softmax(indicator_preds * predictor_confidences, dim=1)
```

Comment:
softmax of an element-wise product of two [0, 1] bounded tensors feels like it should have a temperature

Reply:
does [pytorch](https://pytorch.org/docs/stable/nn.html?highlight=softmax#torch.nn.functional.softmax) include this as a parameter? 

also, do you have some intuition about what $\tau$ to choose?

Reply:
No you'd just add it as an init param to this module then do 

```python
F.softmax(indicator_preds * predictor_confidences / self.tau, dim=1)
```

and I have no idea! how peaky do we want this thing?

Reply:
defaulting to 0.1 for now to make it peakier by default... can surface this as a hyperparameter in applications / tutorials